lab_2_tokenize_by_bpe/main.py

```diff
@@ -26,42 +48,141 @@ def collect_frequencies(
     :param end_of_word: a token that signifies the end of word
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(text, str) or not isinstance(end_of_word, str):
+        return None
+    if not isinstance(start_of_word, str) and start_of_word is not None:
+        return None
+
+    dict_freq = {}
+    words_list = text.split()
+    for word in words_list:
+        prepared_word = prepare_word(word, start_of_word, end_of_word)
+        if prepared_word is None:
+            return None
+        if prepared_word not in dict_freq:
+            dict_freq[prepared_word] = words_list.count(word)
+
+    return dict_freq
 
 
 def count_tokens_pairs(
-    word_frequencies: dict[tuple[str, ...], int]
+        word_frequencies: dict[tuple[str, ...], int]
 ) -> dict[tuple[str, str], int] | None:
     """
     Counts number of occurrences of each pair of subsequent tokens
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+
+    pair_dict = {}
+
+    for word in word_frequencies:
+        for index, token in enumerate(word):
+            if index == len(word) - 1:
+                break
+            next_token_index = index + 1
+            pair = (token, word[next_token_index])
+            pair_dict[pair] = 0
+
+    for pair in pair_dict.copy():
+        freq = 0
+        for word in word_frequencies:
+            for index, token in enumerate(word):
+                if index == len(word) - 1:
+                    break
+                if token == pair[0] and word[index + 1] == pair[1]:
+                    freq += word_frequencies[word]
+                    pair_dict[pair] = freq
+
+    return pair_dict
 
 
 def merge_tokens(
-    word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
+        word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
 ) -> dict[tuple[str, ...], int] | None:
     """
     Updates word frequency dictionary by replacing a pair of token with a merged one
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+
+    word_freq_updated = word_frequencies.copy()
+
+    for word in word_frequencies:
+        new_word = []
+        if pair[0] in word and pair[1] in word:
+            for index, token in enumerate(word):
+                if token == pair[1] and word[index - 1] == pair[0]:
+                    pass
+                elif token != pair[0]:
```

Comment:
so `pair[1]` will be added twice, am I right?

Reply:
no, lines 120 and 121 prevent it

Reply:
Oh they sure do...
![image](https://github.com/fipl-hse/2023-2-level-labs/assets/61208744/ecfa53e9-ca17-4db1-aae6-6afc3e57eec4)

Reply:
This function is hard to understand. Try to implemented in a more clear way.

Reply:
I can't come up with a different way to write this function (